### ISSUE TITLE
`os/sigaction` fails if a separate thread is waiting on `(file/read stdin :line)`.

### ISSUE THREAD

**Issue by amano-kenji** [janet-lang/janet#1378]
Here's the example code.

```janet
(import spork/sh)

(defn action []
  (sh/rm "test")
  (print "Handled SIGINT!")
  (os/exit 1))

(ev/spawn-thread
  (let [server (net/listen :unix "test" :stream)]
    (forever
      (with [conn (net/accept server)]
        (print (ev/read conn :all))))))

(ev/spawn-thread
  (forever
    (print (file/read stdin :line))))

(defn main [_]
  (os/sigaction :int action true)
  (forever
    (ev/sleep 1)))
```

I had to press Ctrl+C multiple times to kill this script.

**Comment by amano-kenji** [comment]
How can I use sigmask?

**Comment by iacore** [comment]
```
man sigprocmask
man sigwait
```

you can mask the selected signals on every thread, and use `sigwait`

**Comment by bakpakin** [comment]
There is definitely some improvement to be made here - we can probably have better behavior by default, and possibly expose sigprocmask for control over which threads handle signals. Regardless, the current behavior isn't great.

**Comment by bakpakin** [comment]
An aside - a quick check on my machine seems to indicate that using `:term` SIGTERM instead of `:int` SIGINT will have the desired effect

**Comment by bakpakin** [comment]
Looking into this some more, I think the issue is some undefined behavior of what happens when you called `exit` in a program with pthread "detached" threads, of which all Janet's threads are. This is certainly an issue that needs some careful consideration - by default, calling exit will run `atexit` registered functions in each thread. One "fix" for this issue is to replace `exit` in janet's source code with `_exit`, which bypasses this atexit functionality that pthread presumably uses for cleanup.

I'm not sure this is an ideal solution, though.

**Comment by iacore** [comment]
> some undefined behavior 

On Linux, the process group is killed together. The `atexit` is a libc feature. 

Does janet use `atexit` at all?

**Comment by iacore** [comment]
I found another issue while playing around with signal handler.

During the signal handler, the process signal mask is not changed (deviation from the POSIX behavior). So if you send SIGINT inside the Janet signal handler, the expected behavior is to not intercept the program.


Current behavior:
Pressing Ctrl+C makes the following program go into a loop.

```
(import spork/sh)

(def kill (ffi/lookup (ffi/native) "kill"))
(def signature (ffi/signature :default :int :int :int))


(defn action []
  (sh/rm "test")
  (print "Handled SIGINT!")
  (ffi/call kill signature 0 2)
  (print "unreachable"))

(ev/spawn-thread
  (let [server (net/listen :unix "test" :stream)]
    (forever
      (with [conn (net/accept server)]
        (print (ev/read conn :all))))))

(ev/spawn-thread
  (forever
    (print (file/read stdin :line))))

(defn main [_]
  (os/sigaction :int action)
  (forever
    (ev/sleep 1)))
```

**Comment by iacore** [comment]
@amano-kenji here's a janky workaround you can use for now

```
(def kill (ffi/lookup (ffi/native) "kill"))
(def signature (ffi/signature :default :int :int :int))


(defn action []
  (sh/rm "test")
  (print "Handled SIGINT!")
  (ffi/call kill signature 0 15))
```

**Comment by bakpakin** [comment]
This has been fixed on the latest master with the addition of a `force` argument to exit that calls `_exit` instead. The cause of this is just that `exit()` flushed stdio streams before exiting, as per the C standard, so this is not an issue with the interpreter.

**Comment by sogaiu** [comment]
So since f6df8ff93 `os/exit` has the docstring:

```
    (os/exit &opt x force)

    Exit from janet with an exit code equal to x. If x is not an 
    integer, the exit with status equal the hash of x. If `force` is 
    truthy will exit immediately and skip cleanup code.
```

**Comment by amano-kenji** [comment]
> skip cleanup code

Does this mean `os/sigaction` will be skippped?

**Comment by amano-kenji** [comment]
I just tested the latest commit.

`(os/exit 1 true)` can be called in signal handlers, but `(os/exit 1)` skips signal handlers.
